gapic/samplegen/samplegen.py

```diff
@@ -0,0 +1,359 @@
+# Copyright (C) 2019  Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import keyword
+import re
+import gapic.samplegen.utils
+
+from collections import defaultdict
+
+# Outstanding issues:
+# * Neither license nor copyright are defined
+# * In real sample configs, many variables are
+#   defined with an _implicit_ $resp variable.
+
+MIN_SCHEMA_VERSION = (1, 2, 0)
+
+VALID_CONFIG_TYPE = 'com.google.api.codegen.SampleConfigProto'
+
+# There are a couple of other names that should be reserved
+# for sample variable assignments, e.g. 'client', but there are diminishing returns:
+# the sample config author isn't a pathological adversary, they're a normal person
+# who may make mistakes occasionally.
+RESERVED_WORDS = frozenset(itertools.chain(keyword.kwlist, dir(__builtins__)))
+
+TEMPLATE_NAME = 'samplegen_template.j2'
+
+
+class SampleError(Exception):
+    pass
+
+
+class ReservedVariableName(SampleError):
+    pass
+
+
+class SchemaVersion(SampleError):
+    pass
+
+
+class RpcMethodNotFound(SampleError):
+    pass
+
+
+class InvalidConfigType(SampleError):
+    pass
+
+
+class InvalidStatement(SampleError):
+    pass
+
+
+class BadLoop(SampleError):
+    pass
+
+
+class MismatchedPrintStatement(SampleError):
+    pass
+
+
+class UndefinedVariableReference(SampleError):
+    pass
+
+
+class BadAssignment(SampleError):
+    pass
+
+
+class InconsistentRequestName(SampleError):
+    pass
+
+
+class InvalidRequestSetup(SampleError):
+    pass
+
+
+class DuplicateInputParameter(SampleError):
+    pass
+
+
+def validate_response(response):
```

Comment:
Using type hints consistently is difficult because the input yaml is, effectively, an AST using lists, dicts, and strs.
I can cheat a little with `validate_and_transform_request`.
Doing the same for response validation, particularly loops because of recursion, would require a class hierarchy and a stronger degree of tie in with the parsing stage. I don't think the added complexity is worth the resulting static analysis help.